File: AbstractFactory/python-source/rooms/BasementHardRoom.py
from random import random

from rooms import Room
from rooms.HardRoom import HardRoom


class BasementHardRoom(HardRoom):
    def generate_loot(self) -> str:
        return 'Loot x2'

    def generate_boss(self) -> str:
        return '200hp'

    def generate_secret(self, previous_room: Room) -> bool:
        if isinstance(previous_room, BasementHardRoom):
            return random() < 0.2
        else:
            return random() < 0.1

File: JavaSource/src/main/java/ru/akhcheck/patterns/abstractfactory/rooms/BasementHardRoom.java
package ru.akhcheck.patterns.abstractfactory.rooms;

/** Сложная комната подвала: у неё есть секрет. */
public class BasementHardRoom implements HardRoom {
    @Override
    public String generateLoot() {
        return "Basement loot x2";
    }

    @Override
    public String generateBoss() {
        return "Monstro 200hp";
    }

    @Override
    public String generateSecret(Room neighbour) {
        return "secret behind " + neighbour.generateBoss();
    }
}
